### PR TITLE
Make the kobuki base actually work

### PR DESCRIPTION
**Pull request by clalancette**
This PR is a series of fixes to make the Kobuki base actually work.  It requires https://github.com/stonier/ecl_core/pull/101 to actually build properly.  With these changes in place, I was able to verify:

1.  That I can command the base by sending a twist to `/command/velocity`.
1.  That the base comes to a halt automatically when the twist stops coming in.
1.  The IMU publishes data.
1.  The buttons publish data.
1.  The cliff sensors publish data.
1.  The wheel drops publish data.
1.  The buttons publish data.
1.  The bumper publishes data.
1.  The dock_ir sensor publishes data.
1.  Odometry data is published.
1.  Tf data is published.
1.  Diagnostic data is published.

I haven't yet tried out some of the other commands (like setting the LEDs, resetting odometry, etc).  I also need to finish porting the launch files and configuration files.  But this is a good start.

### CODE REVIEW

**Comment by clalancette** [comment]
@stonier Now that we have merged and released the fixes for ecl_core stuff, I can now build this package against Eloquent.  We still won't be able to port all of the launch files until we deal with https://github.com/kobuki-base/cmd_vel_mux/issues/1 , but I think this is in good enough state to review and merge this.  Thanks!

**Comment by clalancette** [comment]
And I've now ported the launch files and configuration files to ROS 2.

**Comment by clalancette** [comment]
@stonier Friendly ping here; mind reviewing and merging this if you think it is good?

**Comment by clalancette** [comment]
Another friendly ping; can we go ahead and merge this?  It would be nice to get this released for Eloquent and Foxy.

**Comment by emersonknapp** [comment]
Hello @stonier ! I'd like these changes, too - just bumping your notification stream :)

File: kobuki_node/src/odometry.cpp [hunk 1]
```diff
@@ -55,8 +66,14 @@ bool Odometry::commandTimeout(const rclcpp::Time & now) const {
   }
 }
 
+void Odometry::resetTimeout(const rclcpp::Time & now)
+{
+  last_cmd_time_ = now;
```
Comment (stonier) [inline kobuki_node/src/odometry.cpp:71]: General migration of inline code to source code? We used to be in the habit of optimising on arm cores any one-liner to an inline call. Not really an issue for kobuki though, just wondering if there was a significant reason you do so here?

File: kobuki_node/src/odometry.cpp [hunk 2]
```diff
@@ -14,12 +14,12 @@
...
 #include <string>
 
-#include <geometry_msgs/msg/quaternion.hpp>
 #include <geometry_msgs/msg/transform_stamped.hpp>
 #include <nav_msgs/msg/odometry.hpp>
+#include <rclcpp/rclcpp.hpp>
 #include <tf2/LinearMath/Quaternion.h>
 
-#include <ecl/geometry.hpp>
+#include <ecl/geometry/legacy_pose2d.hpp>
```
Comment (stonier) [inline kobuki_node/src/odometry.cpp:22]: Need to deprecate this. Don't need to do it here though.

https://github.com/kobuki-base/kobuki_ros/issues/11

File: kobuki_node/src/kobuki_ros.cpp [hunk 3]
```diff
@@ -948,6 +946,12 @@ void KobukiRos::publishRawDataCommand(kobuki::Command::Buffer &buffer)
  */
 void KobukiRos::publishRawDataStream(kobuki::PacketFinder::BufferType &buffer)
 {
+  if (raw_data_stream_publisher_->get_subscription_count() == 0 &&
+    raw_data_stream_publisher_->get_intra_process_subscription_count() == 0)   // no one listening?
```
Comment (stonier) [inline kobuki_node/src/kobuki_ros.cpp:950]: This api is a little suprising. I would have inferred that `get_subscription_count` was the sum total of inter and intra subscriptions.

File: kobuki_node/src/kobuki_ros.cpp [hunk 4]
```diff
@@ -648,14 +640,14 @@ void KobukiRos::publishInertia()
 void KobukiRos::publishRawInertia()
 {
     // Publish as unique pointer to leverage the zero-copy pub/sub feature
-  auto msg = std::make_unique<sensor_msgs::msg::Imu>();
   kobuki::ThreeAxisGyro::Data data = kobuki_.getRawInertiaData();
 
   rclcpp::Time now = this->get_clock()->now();
-  rclcpp::Duration interval(0.01); // Time interval between each sensor reading.
+  rclcpp::Duration interval(RCL_S_TO_NS(0.01)); // Time interval between each sensor reading.
```
Comment (stonier) [inline kobuki_node/src/kobuki_ros.cpp:646]: Oh, that's a surprise. Good catch. I'll keep an eye out elsewhere.

File: kobuki_node/src/diagnostics.cpp [hunk 5]
```diff
@@ -26,143 +28,167 @@ namespace kobuki_node
...
       stat.summary(diagnostic_msgs::msg::DiagnosticStatus::OK, "Healthy");
       break;
-    }
-    case ( kobuki::Battery::Low ) : {
+    case kobuki::Battery::Low:
       stat.summary(diagnostic_msgs::msg::DiagnosticStatus::WARN, "Low");
       break;
-    }
-    case ( kobuki::Battery::Dangerous ) : {
+    case kobuki::Battery::Dangerous:
```
Comment (stonier) [inline kobuki_node/src/diagnostics.cpp:43]: Never liked this much. Technically correct, but cognitively expensive given that it looks like separate code blocks and the emphasis to put curly scope around every other code block. Always seems to lead to raised eyebrows and problems for people the first time they hit them.

**Comment by stonier** [comment]
@clalancette Perhaps needed an unfriendly ping! Still catching up elsewhere and with the underlying kobuki repositories after a bit of a haitus. Lockdown and remote working was fun, software testing actually became imperative. However, am finally ... moving ... again ... on kobuki! :)

I haven't tested this, but looks like you've done some preliminary testing and code looks to be in good shape to be a useful launching pad for hardening the progress from here. We were in a non-functional state anyway, so this is a big step :)

Just left some comments / queries, but let's get this in.

**Comment by stonier** [comment]
Squashed with your useful PR description as the commit summary.

File: kobuki_node/src/odometry.cpp (hunk 1 above)
Comment (clalancette) [reply kobuki_node/src/odometry.cpp:71]: Just consistencies sake; for me, it is easier to just read the C++ file for implementation and not have to open the header file unless I'm changing signatures.

File: kobuki_node/src/odometry.cpp (hunk 2 above)
Comment (clalancette) [reply kobuki_node/src/odometry.cpp:22]: I'll take a look.

File: kobuki_node/src/kobuki_ros.cpp (hunk 3 above)
Comment (clalancette) [reply kobuki_node/src/kobuki_ros.cpp:950]: Yeah, its a crufty part of the rclcpp API that I keep meaning to change, but have never gotten around to.

**Comment by clalancette** [comment]
> @clalancette Perhaps needed an unfriendly ping! Still catching up elsewhere and with the underlying kobuki repositories after a bit of a haitus. Lockdown and remote working was fun, software testing actually became imperative. However, am finally ... moving ... again ... on kobuki! :)

No worries!  Thanks for taking a look.

> I haven't tested this, but looks like you've done some preliminary testing and code looks to be in good shape to be a useful launching pad for hardening the progress from here. We were in a non-functional state anyway, so this is a big step :)

I've been running the robot around (on Eloquent) with this patch for a while now, and things seem to be working OK for me.  So we're at least in a semi-functional state, though I can't say I've tested every combination of everything.  I'll take a look at the PR for Foxy as well now.

Thanks again!

**Comment by emersonknapp** [comment]
I was running around using this patch (+ a tiny fix for std::move semantics) on Foxy this weekend

**Pull request merged**